notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/InstallationPutRequest.java

```diff
@@ -0,0 +1,192 @@
+            }};
+            return params;
+        } catch (InvalidKeyException e) {
+            e.printStackTrace();
+            return null;
+        }
+    }
+
+    @Override
+    protected Response<JSONObject> parseNetworkResponse(NetworkResponse response) {
+
+        if(response.statusCode == HttpURLConnection.HTTP_OK) {
+            return Response.success(new JSONObject(), HttpHeaderParser.parseCacheHeaders(response));
+        }
+
+        return super.parseNetworkResponse(response);
+    }
+
+    public static JSONObject getBody(final Installation installation) {
+        final JSONArray tagList = new JSONArray();
+        for (String tag: installation.getTags()) {
+            tagList.put(tag);
+        }
+
+        try {
+            final JSONObject serializedTemplates = new JSONObject();
+            for (Map.Entry<String, InstallationTemplate> template : installation.getTemplates().entrySet()) {
+                String templateName = template.getKey();
+                serializedTemplates.put(templateName, InstallationTemplate.serialize(templateName, template.getValue()));
+            }
+
+            JSONObject jsonBody = new JSONObject() {{
+                put("installationId", installation.getInstallationId());
+                put("platform", "GCM");
+                put("pushChannel", installation.getPushChannel());
+                put("tags", tagList);
+                put("templates", serializedTemplates);
+            }};
+
+            Date expiration = installation.getExpiration();
+            if (expiration != null) {
+                String formattedExpiration = sIso8601Format.format(expiration);
+                jsonBody.put("expirationTime", formattedExpiration);
+            }
+            return jsonBody;
+        } catch (JSONException e) {
+            // Converting to a RuntimeException, avoiding specific checking which would break
+            throw new UnsupportedOperationException("", e);
+        }
+    }
+
+    public static <T> Response.Listener<T> wrapListener(final InstallationAdapter.Listener subject, final Installation installation) {
+        return new Response.Listener<T>() {
+            @Override
+            public void onResponse(T response) {
+                subject.onInstallationSaved(installation);
+            }
+        };
+    }
+
+    public static Response.ErrorListener wrapErrorListener(final InstallationAdapter.ErrorListener errorListener) {
+        return new Response.ErrorListener() {
+            @Override
+            public void onErrorResponse(VolleyError error) {
+                errorListener.onInstallationSaveError(error);
+            }
+        };
+    }
+
+    /**
+     * Generates the User-Agent
+     */
+    private static String getUserAgent() {
+        String userAgent = String.format("NOTIFICATIONHUBS/%s (api-origin=%s; os=%s; os_version=%s;)",
+                API_VERSION, "AndroidSdkV1FcmV1.0.0-preview3", "Android", Build.VERSION.RELEASE);

```

Comment:
Should this be here for GA or read from somewhere else?

Reply:
Fixed in dd81e3708c279fbb07da822a74f002394f94e96a

Reply:
This fix ended up breaking javadoc generation in a sorta tricky way. I'm pulling out the fix and will file a separate bug to follow up with this.

Reply:
We still need to update the version number here as it is still preview

Reply:
Yep, that will come in a separate PR.